src/Database/Concerns/HasRelationships.php

```diff
@@ -279,520 +338,573 @@ protected function getRelationDefaults($type)
+        $relation->setRelationName($relationName);
+
+        // Add dependency, if required
+        if (
+            in_array(
+                \Winter\Storm\Database\Relations\Concerns\CanBeDependent::class,
+                class_uses_recursive($relation)
+            )
+            && (($definition['delete'] ?? false) === true)
+        ) {
+            $relation = $relation->dependent();
+        }
 
-        // Query filter arguments
-        $filters = ['scope', 'conditions', 'order', 'pivot', 'timestamps', 'push', 'count', 'default'];
+        // Remove detachable, if required
+        if (
+            in_array(
+                \Winter\Storm\Database\Relations\Concerns\CanBeDetachable::class,
+                class_uses_recursive($relation)
+            )
+            && (($definition['detach'] ?? true) === false)
+        ) {
+            $relation = $relation->notDetachable();
+        }
 
-        foreach (array_merge($optional, $filters) as $key) {
-            if (!array_key_exists($key, $relation)) {
-                $relation[$key] = null;
-            }
+        // Remove pushable flag, if required
+        if (
+            in_array(
+                \Winter\Storm\Database\Relations\Concerns\CanBePushed::class,
+                class_uses_recursive($relation)
+            )
+            && (($definition['push'] ?? true) === false)
+        ) {
+            $relation = $relation->noPush();
         }
 
-        $missingRequired = [];
-        foreach ($required as $key) {
-            if (!array_key_exists($key, $relation)) {
-                $missingRequired[] = $key;
-            }
+        // Add count only flag, if required
+        if (
+            in_array(
+                \Winter\Storm\Database\Relations\Concerns\CanBeCounted::class,
+                class_uses_recursive($relation)
+            )
+            && (($definition['count'] ?? false) === true)
+            && $addConstraints
+        ) {
+            $relation = $relation->countOnly();
         }
 
-        if ($missingRequired) {
-            throw new InvalidArgumentException(sprintf(
-                'Relation "%s" on model "%s" should contain the following key(s): %s',
-                $relationName,
-                get_called_class(),
-                implode(', ', $missingRequired)
-            ));
+        $relation->addDefinedConstraintsToRelation();
+
+        if ($addConstraints) {
+            // Add defined constraints
+            $relation->addDefinedConstraintsToQuery();
         }
 
         return $relation;
     }
 
     /**
-     * Define a one-to-one relationship.
-     * This code is a duplicate of Eloquent but uses a Storm relation class.
-     * @return \Winter\Storm\Database\Relations\HasOne
+     * Finds the calling function name from the stack trace.
      */
-    public function hasOne($related, $primaryKey = null, $localKey = null, $relationName = null)
+    protected function getRelationCaller(): ?string
```

Comment:
What is this being used for?

Reply:
It's used to determine the relation name for relations created by calling a method when you've only defined property-style relations. It was in the original version of this file.

Basically, it's so, for example:

```
$comments = $post->comments()->get();
```

When calling `comments()`, if that was defined as a property relation, it uses `comments` as the relation name when creating the relation object, and determining which config to retrieve.